src/gluonts/shell/sagemaker/dyn.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
+`gluonts.shell.sagemaker.dyn`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Dynamic code loading for `gluonts.shell`.
+"""
+
+
+import os
+import shutil
+import subprocess
+import sys
+import tarfile
+import tempfile
+from functools import partial
+from pathlib import Path
+from typing import Optional
+
+
+class Installer:
+    def __init__(self, packages):
+        self.packages = packages
+        self.cleanups = []
+
+    def cleanup(self):
+        for cleanup in self.cleanups:
+            cleanup()
+
+    def copy_install(self, path: Path):
+        if path.is_file():
+            shutil.copy(path, self.packages)
+        elif path.is_dir():
+            shutil.copytree(path, self.packages / path.name)
+
+    def pip_install(self, path: Path):
+        subprocess.check_call(
+            [
+                sys.executable,
+                "-m",
+                "pip",
+                "install",
+                "--upgrade",
+                "--target",
+                str(self.packages),
+                str(path),
+            ]
+        )
+
+    def install(self, path):
+        if path.is_file():
+            if path.suffix == ".py":
+                self.copy_install(path)
+
+            elif path.suffixes[-2:] == [".tar", ".gz"]:
+                with tarfile.open(path, "r:gz") as archive:
```

Comment:
Is the package automatically tar zipped by the shell or can the user do it himself?

Any reason to hardcode gzip-compression here?
If user is allowed to upload any tarfile, you might use `elif tarfile.is_tarfile(path):` here. That allows other suffixes (like .tgz) and more compression options (bz, bz2, ..).

Reply:
Thanks for the suggestion. The initial idea was to support `.tar.gz` files which are emitted by `python setup.py sdist`. These are pip-installable directly.

However, I didn't want to blankly install any archive, and thus I opted for unpacking and treating the content normally.

I guess we could also think about supporting `.zip`, which we are at it.